src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs

```diff
@@ -1227,38 +1227,130 @@ protected static CSharpCompilation CreateCompilationWithMscorlibAndSpanSrc(strin
 
         private static string spanSource = @"
 namespace System
-{
-    public readonly ref struct Span<T> 
     {
-        public ref T this[int i] => throw null;
-        public override int GetHashCode() => 1;
-        public int Length { get; }
-
-        unsafe public Span(void* pointer, int length)
+        public readonly ref struct Span<T>
         {
-            this.Length = length;
+            private readonly T[] arr;
+
+            public ref T this[int i] => ref arr[i];
+            public override int GetHashCode() => 1;
+            public int Length { get; }
+
+            unsafe public Span(void* pointer, int length)
+            {
+                this.arr = null;
+                this.Length = length;
+            }
+
+            public Span(T[] arr)
+            {
+                this.arr = arr;
+                this.Length = arr.Length;
+            }
+
+            public void CopyTo(Span<T> other) { }
+
+            /// <summary>Gets an enumerator for this span.</summary>
+            public Enumerator GetEnumerator() => new Enumerator(this);
```

Comment:
Does the enumerator matter anymore? Is it exercised?

Reply:
I just copied it from the actual implementation. Anyone can call it. It also works as a fallback, in case we do have to emit 'foreach'

---
In reply to: [153598460](https://github.com/dotnet/roslyn/pull/23363#discussion_r153598460) [](ancestors = 153598460)

Reply:
@VSadov, ideally we wouldn't need the Enumerator.  With your PR, can I get rid of it from the type?

Reply:
This PR does not introduce any new C# features.  
It is basically an optimization for a special case where we can emit better code. 

If type does not have `GetEnumerator` of a particular shape, it would not be considered as something that can be used with `foreach`.